Use host PID namespace in docker example

snarlysodboxer opened the pull request:
See https://github.com/prometheus/node_exporter/issues/671

I'm unclear as to if the container needs the `--privileged` option.

Mentioning @SuperQ and @discordianfish according to the contributing guidelines.

SuperQ commented on README.md:
```diff
@@ -122,28 +122,11 @@ To see all available configuration flags:
-    --collector.procfs /host/proc \
-    --collector.sysfs /host/sys \
-    --collector.filesystem.ignored-mount-points "^/(sys|proc|dev|host|etc)($|/)"
+docker run -d --net="host" --pid="host" quay.io/prometheus/node-exporter
```
Please keep the multi-line style for readability.

discordianfish commented:
Good point! But also not sure why you removed the explaination around it.
And you're right, it probably doesn't need --privileged either. Somebody needs to test that though.

snarlysodboxer commented:
Sorry for the delay guys, I have updated the PR.

@discordianfish Is is still "not recommended" to run node_exporter in Docker when you can give it the host's PID and Network Namespaces? - Why would it matter at that point? - That's why I removed the explanation, I figured it was now unnecessary.

discordianfish commented:
You still need to bind-mount all mountpoints and even then you end up with 'ugly' filesystem labels (like /host/proc). There might be even more limitation.. would be good to find out exactly but in general the prefered way is not to run it under docker.

snarlysodboxer commented:
@discordianfish I might be missing something here, but I understand that giving the container the host's PID namespace nullifies the need for bind mounting anything into the container.

As mentioned in my [initial comment](https://github.com/prometheus/node_exporter/issues/671#issue-255100754) about this, there are significant container-centric communities (kubernetes, coreos) running node_exporter without bind mounting anything, - because of the host PID shared namespace.

discordianfish commented:
> @discordianfish I might be missing something here, but I understand that giving the container the host's PID namespace nullifies the need for bind mounting anything into the container.

I don't think so. Have you tried? That's easy to verify.

> As mentioned in my initial comment about this, there are significant container-centric communities (kubernetes, coreos) successfully running node_exporter without bind mounting anything, - because of the host PID shared namespace.

I'm aware of that, they do so because they have no other way to deploy the node_exporter. I wish there would be a better way but best option is still to run the exporter on the host directly.

discordianfish commented:
(fwiw, the node-exporter chart also bind-mounts: https://github.com/kubernetes/charts/blob/master/stable/prometheus/templates/node-exporter-daemonset.yaml#L63)

snarlysodboxer commented:
@discordianfish The whole reason I'm making this pull request is because not only have I successfully* tried it, but large communities of peoples have tried and are apparently successfully using node_exporter in Docker without any bind mounting.

*Everything that I would expect to work works, and the only reason I can't judge repletely for us all  (which is what I feel like I'm being asked to do when you say "easy to verify") is because I'm not being paid to dev on node_exporter and run time-based side by side comparisons inside and outside of docker. I thought your team would be the experts on that, especially since Prometheus 2.0's purpose is trying to solve the types of problems that container orchestration systems' high cardinality often present, as seen in for example Kubernetes. 

I'm trying to help you guys be more accessible, and when me and my peers see "not recommended for containerization" it's an orchestration complication at least, if not a design smell. (Plus, who want's to build a special host machine JUST so they can run node_exporter directly on it, when otherwise we can use any generic linux host machine.)

I have strong suspicions about how it's being done, but I'm not absolutely positive so I don't want to speak as an authority, but if I understand correctly, sharing a PID namespace effectively nullifies the need for bind mounting because it's **not limiting the process (aka 'container')** to separate (than the host) `/` `/proc` and `/sys` filesystems.

discordianfish commented:
>  I'm not being paid to dev on node_exporter

Me neither.

> I have strong suspicions about how it's being done, but I'm not absolutely positive so I don't want to speak as an authority, but if I understand correctly, sharing a PID namespace effectively nullifies the need for bind mounting because it's not limiting the process (aka 'container') to separate (than the host) / /proc and /sys filesystems.

For sys/procfs but not for other mountpoints:

```
$ curl localhost:9100/metrics |sed 's/ [0-9\.e\+]*$//' > ne.docker; curl localhost:9101/metrics |sed 's/ [0-9\.e\+]*$//' > ne.host
$ diff -u ne.docker ne.host |grep node_filesystem_free
 # HELP node_filesystem_free Filesystem free space in bytes.
 # TYPE node_filesystem_free gauge
-node_filesystem_free{device="/dev/mapper/nvme0n1p7_crypt",fstype="ext4",mountpoint="/etc/hostname"}
-node_filesystem_free{device="/dev/mapper/nvme0n1p7_crypt",fstype="ext4",mountpoint="/etc/hosts"}
-node_filesystem_free{device="/dev/mapper/nvme0n1p7_crypt",fstype="ext4",mountpoint="/etc/resolv.conf"}
-node_filesystem_free{device="overlay",fstype="overlay",mountpoint="/"}
+node_filesystem_free{device="/dev/mapper/nvme0n1p7_crypt",fstype="ext4",mountpoint="/"}
+node_filesystem_free{device="/dev/mapper/nvme0n1p7_crypt",fstype="ext4",mountpoint="/var/lib/docker/overlay2"}
+node_filesystem_free{device="/dev/nvme0n1p1",fstype="vfat",mountpoint="/boot/efi"}
+node_filesystem_free{device="/dev/nvme0n1p6",fstype="ext4",mountpoint="/boot"}
+node_filesystem_free{device="gvfsd-fuse",fstype="fuse.gvfsd-fuse",mountpoint="/run/user/1000/gvfs"}
+node_filesystem_free{device="sunrpc",fstype="rpc_pipefs",mountpoint="/run/rpc_pipefs"}
+node_filesystem_free{device="tmpfs",fstype="tmpfs",mountpoint="/run"}
+node_filesystem_free{device="tmpfs",fstype="tmpfs",mountpoint="/run/lock"}
+node_filesystem_free{device="tmpfs",fstype="tmpfs",mountpoint="/run/user/1000"}
```

Please do not remove the note and we can merge this PR.

SuperQ commented:
@snarlysodboxer

> I'm not being paid to dev

Nobody is, the node_exporter is an entirely community driven project.

>  when me and my peers see "not recommended for containerization" it's an orchestration complication at least, if not a design smell. (Plus, who want's to build a special host machine JUST so they can run node_exporter directly on it, when otherwise we can use any generic linux host machine that runs Docker.)

This is not a "design smell", this is a intentional design.  The node_exporter is system software, much like init, sshd, etc.  It is required to run in the system root container, as it is providing exporting of metrics for the whole system.  You're conflating orchestration, configuration management, and containerization.  Do you consider it "building a special host machine" to include sshd?

The problem here is Docker.  It bundles software packaging and runtime, but does not provide a runtime abstraction that makes it possible to easily run the node_exporter in a way that makes sense.  Contrast that to the [rkt fly stage1](https://coreos.com/rkt/docs/latest/running-fly-stage1.html) which provides the required environment for the node_exporter to run in.

snarlysodboxer commented:
> Nobody is, the node_exporter is an entirely community driven project.

Doh, my bad. We really appreciate your work!

> This is not a "design smell", this is a intentional design.

I doubt that any of us are "intentionally" designing node_exporter to require things that container runtimes cannot support just to spite people who use them. I do understand that a side effect of node_exporter's goals/needs may be that Docker won't support them all. 

> You're conflating orchestration, configuration management, and containerization. Do you consider it "building a special host machine" to include sshd?

You're right, I was conflating them. But again, coming from an accessibility standpoint, cloud providers provide machine images with SSHd and even Docker installed, but for understandable reasons, not node_exporter.

We can run large clusters with hundreds of apps without customizing a machine image _or_ doing any configuration management, but if we want to use node_exporter, we have to diverge from the machine upgrade path and build and future-support custom machine images just to run node_exporter, and that makes us want to start looking for alternatives to node_exporter. :-(

> The problem here is Docker. It bundles software packaging and runtime, but does not provide a runtime abstraction that makes it possible to easily run the node_exporter in a way that makes sense. Contrast that to the rkt fly stage1 which provides the required environment for the node_exporter to run in.

Agreed!

For those of us using Kubernetes, Docker works perfectly because we _don't want_ the data on non-root mount points anyways, as those network disks are frequently moved around between the host machines making tracking them useless, and we have other ways of tracking them anyway.

discordianfish commented:
What are we discussing here right now? This is a PR to change the README. We said we should keep the recommendation to run it straight on the host because of the mentioned downsides of using host pid/network namespace. The updated PR is better and I'd be fine with merging it, althought we should probably add a sentence why it's not recommended to avoid further confusion. @SuperQ wdyt?

> We can run large clusters with hundreds of apps without customizing a machine image or doing any configuration management, but if we want to use node_exporter, we have to diverge from the machine upgrade path and build and future-support custom machine images just to run node_exporter, and that makes us want to start looking for alternatives to node_exporter. :-(

You don't have to, if you understand the implications it's perfectly fine to run it on docker. It's up to you. It's just not recommended for the reasons stated here already. I think we can all agree on these, right? You are welcome to fix all these downsides of running in docker and we might change this recommendation.

discordianfish commented:
Lets merge this, we can tweak the readme later.
